### ISSUE TITLE
[Community permissions] Find a way to mitigate users buying a wallet signature "attack"

### ISSUE THREAD

**Issue by jrainville** [status-im/status-app#10094]
# Description

As discussed in the meeting, it seems like it would be possible for a user that doesn't have access to a community, because they do not possess the required tokens, to just buy/get a wallet signature from someone who does.

Since the current way to create the signature is that we sign the address with the wallet account, and then sign it with the Status account, thus linking the two, someone could actually cheat and get someone else to create the address wallet signature, get that signature and then sigh it with their own Status account. That would create a fake link between those two.

We need to find a way to either limit that or remove that possibility altogether.

This ticket is to first investigate a solution and then either close and fix in another issue or fix using this ticket.

**Comment by jrainville** [comment]
Some brainstorm ideas, let me know if I'm saying anything that is not doable:

1. Add a timestamp to both signatures and the timestamps need to be within X seconds for it to be valid (let's say 5 seconds?)
    - Maybe it's still possible to fake the timestamp inside the wallet address signature, but it does seem harder to get  a "fake" wallet signature without the "attackers" having an actual software to do it for them.
2. Find a way to sign at the same time with both keys
    - I have no idea if this exists, but some smart cryptography expert probably has thought of something like that before
    - Basically, if we had a signature method that takes both keys at the same time to create a single signature, it would solve the problem.
    - I'm not sure if it's safe or doable to do on our side, but since status-go has access to all keystores, it should be doable.
3. What was proposed was to limit the number of times an address can be used (validated by the owner)
    - This is not great, because it's very possible that someone would have multiple Status accounts and import the same wallet address in them, so we can't just block the re-using of the same address, so any limit we use would be arbitrary
    - It still would work to mitigate the problem slightly

**Comment by osmaczko** [comment]
Interesting problem.

> 2. Find a way to sign at the same time with both keys
>    
>    * I have no idea if this exists, but some smart cryptography expert probably has thought of something like that before
>    * Basically, if we had a signature method that takes both keys at the same time to create a single signature, it would solve the problem.
>    * I'm not sure if it's safe or doable to do on our side, but since status-go has access to all keystores, it should be doable.

This would be best. I did quick research and found this: https://github.com/hbakhtiyor/schnorr/#aggregatesignaturesprivatekeys-bigint-message-32byte-64byte-error. Although, I am not sure [how it works with secp256k1 keys](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki).

![image](https://user-images.githubusercontent.com/33099791/229090479-6f86b9c7-2b80-4f12-b815-120863405875.png)

**Comment by caybro** [comment]
> Interesting problem.
> 
> > 2. Find a way to sign at the same time with both keys
> >    
> >    * I have no idea if this exists, but some smart cryptography expert probably has thought of something like that before
> >    * Basically, if we had a signature method that takes both keys at the same time to create a single signature, it would solve the problem.
> >    * I'm not sure if it's safe or doable to do on our side, but since status-go has access to all keystores, it should be doable.
> 
> This would be best. I did quick research and found this: https://github.com/hbakhtiyor/schnorr/#aggregatesignaturesprivatekeys-bigint-message-32byte-64byte-error. Although, I am not sure [how it works with secp256k1 keys](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki).

The very first line says:
> This is a Go implementation of the standard 64-byte Schnorr signature scheme over the elliptic curve secp256k1.

**Comment by jrainville** [comment]
I read the doc on that library and it seems like what we need. It would totally fix the attack vector we have. I'll wait to start implementation on this until more people can think it through and see if there are maybe edge cases we didn't consider. Also, it's not necessary for the RC.

**Comment by saledjenic** [comment]
For this issue (when a user wants to join the community and we need to be sure that the signature cannot be reused) we should do the following:

- the user asks the community to join
- community creates a unique message M1 = "CommunityId|UserPublicKey|CurrentTimestamp" (basically anything that ensures uniqueness) that needs to be signed and stores that message M1 in the local table for the key that can be user's address
- the user signs the message M1, received from the community 
- the user sends the final request to join the community
- the community validates the signature and verifies the message comparing it with the one stored in the table and checks if the timestamp is not older than 5 minutes (let's say 5 minutes is the time after which a generated request expires)
- if all went well, the user becomes a member of the community and the request is removed from the table, otherwise error is returned and the request is removed from the table (also we should periodically check and remove expired requests from the table)

Using that approach there is no chance to have user A that will use the signature of user B to successfully join the community.

**Comment by osmaczko** [comment]
> Using that approach there is no chance to have user A that will use the signature of user B to successfully join the community.

Isn't there still a 5-minute window where someone else could provide a signature?

> the user asks the community to join
> community creates a unique message M1 = "CommunityId|UserPublicKey|CurrentTimestamp" (basically anything that ensures uniqueness) that needs to be signed and stores that message M1 in the local table for the key that can be user's address

This process will make joining a community slightly more complex and potentially less reliable in case any message is lost.
```
request to join ->
<- unique message to sign
signed message ->
<- request to join acceptance
```
vs
```
request to join ->
<- request to join acceptance
```

@saledjenic I don't see the advantage of this proposed approach over Schnorr signatures, am I missing something?

**Comment by saledjenic** [comment]
@osmaczko ahh I see above, that you shared the link to Schnorr signatures, well as I see it signs the message with multiple addresses, that's good, the same as what I said, but applying that without making a unique new message that needs to be signed for each new join community request, even for the same user, won't help. Cause at the end even if we use Schnorr signatures over the same message the signature will be the same and somebody else can reuse it, and that's what we're trying to solve. 

In the approach I shared above, when a community generates for each joining community request a new unique message that needs to be signed no way of reusing the signature.

**Comment by osmaczko** [comment]
> Even if we go without a 5-minute window, there is no way that somebody else may sign a message instead of one the message is meant to, until one steals the primary key. A community knows which message is prepared for whom. The only one who can sign it is the one the message is meant to, the community will verify that by inspecting the signature. So if someone else signs the message, the signature will not be correct.

Not sure I understand. Will this message be signed by chat key or wallet? How do we prove we are indeed the owners of given addresses?

> Cause at the end even if we use Schnorr signatures over the same message the signature will be the same and somebody else can reuse it, and that's what we're trying to solve.

The whole idea with Schnorr is to exactly avoid what you wrote. Let me explain it more explicitly. We now send `RequestToJoin` with signed revealed addresses as follows:
```
RequestToJoin + signature(chat key)
 - address1 + signature(wallet key 1)
 - address2 + signature(wallet key 2)
 ...
 - addressN + signature(wallet key N)
```

The problem with the above, is that `addressN + signature(wallet key N)` can be provided by someone else. For instance, if I own tokens that allows me to join certain community, I can send you this piece of data. Then you can put it inside `RequestToJoin` and sign it with your chat key, effectively making yourself being able to join the community.

To solve that, we need to prove we poses all the keys at the same time, that's when Schnorr kicks in, it will look as follows:
```
RequestToJoin + signature(chat key)
 - address1 + signature(wallet key 1 + chat key) aka Schnorr
 - address2 + signature(wallet key 2 + chat key) aka Schnorr
 ...
 - addressN + signature(wallet key N + chat key) aka Schnorr
```

In other words, somebody else can't reuse it, because we use the combination of wallet key and chat key.

**Comment by saledjenic** [comment]
@osmaczko not sure what `chat key` and `wallet key` in your examples refer to? Is that chat public key or chat private key?
Also in that example address is not needed since it can be determined from the signature.
In order to say more, could you tell me what are all the details we're sending when sending a request to join the community?

**Comment by osmaczko** [comment]
> > In order to create a signature you always need a private key, so those are private keys.
> 
> Exactly, but in the example above you said this...
> 
> ```
> RequestToJoin + signature(chat key)
>  - address1 + signature(wallet key 1 + chat key) aka Schnorr
>  - address2 + signature(wallet key 2 + chat key) aka Schnorr
>  ...
>  - addressN + signature(wallet key N + chat key) aka Schnorr
> ```
> 
> Does that mean we need to make a signature of private key of each address?

chat key - is private key of the chat
wallet key - is private key of the wallet

We need to prove (sign) each revealed wallet address with both private keys **at the same time** - the private key of the wallet and private key of the chat key. Schnorr enables that. In pseudocode, again very simplified:

Sender side:
```
  data := addressDataToSign(wallet.address)
  privateKeys := {messenger.privateKey, wallet.privateKey}
  signature := schnorr.sign(privateKeys, data)
  sendMessage(data, signature)
```
Receiver side:
```
  user := getUser(message)
  address := getRevealedAddress(message)
  
  if (!schnorr.verify(user, message.data, message.signature)) {
     error("user didn't sign the message")
  }
  if (!schnorr.verify(address, message.data, message.signature)) {
     error("wallet didn't sign the message")
  }

```

**Comment by saledjenic** [comment]
@osmaczko and I had a call to discuss what's the best way to overcome this issue, considering all the features we have implemented so far.

Conclusion:
- The only real proof that the PublicChat key that is going to join a community is the owner of the addresses that it is going to join with is signing a message combining PrivateKey of the chat key and PrivateKey of the address, and making such signature for each address it wants to reveal to a community. That's possible using Schnorr signatures, need to implement this on the status-go side.
- When we come to a Keycard:
  - We need to extend the library implementing flow that will make a combined signature (using PrivateKey of the chat key and PrivateKey of the address) of the message being signed. This will be used in the case when the revealed address belongs to the profile keypair, _**which means everything will be evaluated on the Keycard side**_. Example of the call `signCombined(pin, address_path, message)`.
  - Now the case when the profile keypair is migrated to a Keycard and the address that needs to be revealed to a community belongs to a keypair that is migrated to another Keycard. There are 2 options here:
    - If we want a private key to never leave the Keycard:
      - then we cannot solve this issue, cause we cannot get the private key of the address to be passed as a param to the profile Keycard to be used for making a combined signature.
    - If it's acceptable for a private key of an address to leave the Keycard:
      - then we need to extend the current `ExportPublic` flow, exporting a private key for the address path, and use that private key as a parameter for running sign combined flow on the profile Keycard. Example of the call `signCombined(pin, private_key, message)`. In this case the only private key that leaves the Keycard is the one related to the wallet account, not one related to the chat.

_Note_:
All that Keycard game can be solved easier if we would collect all private keys of all involved keycards (based on selected addresses + profile keypair state) and send them to status-go, where we will have a function making Schnorr signature, but that decreases security and defeats a bit the purpose of using a Keycard. That's why is maybe better above described than this.

**Comment by bitgamma** [comment]
@saledjenic Keycard does not support Schnorr signatures

**Comment by bitgamma** [comment]
I don't know all the details so correct me if I am wrong. If the current implementation is:

sign(sign(address1, wallet_key), chat_key)

and the proposed solution is

sign(address1, wallet_key + chat_key) => Schnorr

to avoid using Schnorr couldn't we do

sign(sign(address1 + chat_address), wallet_key), chat_key) 

the advantage would be that we retain compatibility with Keycard and unless this violates some requirements I am not aware of, it should solve the same issue since if you do the second sign step with the wrong chat key it would be easy to detect.

**Comment by saledjenic** [comment]
> sign(sign(address1 + chat_address), wallet_key), chat_key)

If wallet_key and chat_key refer to private keys, do we really want to send the message containing private keys over the wire?

If they don't refer to private keys, then it doesn't work, a user will still be able to cheat.

**Comment by saledjenic** [comment]
Let me give you the context about the problem we have. Imagine there is a Community that set a rule saying that only members who have more than 10 ETH on their accounts can join. 

Now imagine there is a poor user A (has less than 10 ETH) who wants to join the Community and a rich user B (has 10+ ETH).

Now, a joining community request among other things contains the signature of the message signed by chat_private_key and the signature of the message signed by the wallet_private_key. That's how it's implemented now and how we prove that the owner of the chat key is the same as the owner of the wallet address. That approach is not good cause nothing stops user A from paying significantly less to user B to sign the message for him and user A can use that signature to prove that he owns the wallet address and the Community has no way to check that.

What you're saying is:
1. sign message composed of `WalletAddress + ChatAddress` using wallet_private_key
2. sign the signature made in point 1. using chat_private_key

That doesn't work, cause user A can still ask user B to sign the message `WalletAddress (of user B) + ChatAddress (of user A)` and send him back the signature which user A can sign using his chat_private_key and send the request to join the Community.

So as far as we see we need to include both private keys (chat_private_key + wallet_private_key) in signing the message/request to join.

**Comment by John-44** [comment]
FYI @iurimatias for visibility

**Comment by bitgamma** [comment]
I was missing the point that user A and user B are collaborating in the attack. I was under the assumption user B is being attacked by user A. 

If user A and B are collaborating, Schnorr signatures do not help. User A and User B can each produce a signature with just their key and then user A can aggregate them by simply summing them. More info at  https://blog.blockstream.com/en-musig-key-aggregation-schnorr-signatures/. The idea is that individual signatures can be aggregated into a combined one after being generated

**Comment by guylouis** [comment]
anyone with a view on how to solve this then?

**Comment by guylouis** [comment]
for info I quickly mentioned this problem with @cammellos, and here's my understanding of the discussion:
- there might not be cryptographic ways to prevent this: a user A could always collaborate with a user B to provide some information that B would sign with his wallet account. Thus B is effectively "lending" a wallet account to A. 
- if we want to prevent this, which is probably a product decision (@John-44 ) then an option would be to have the community admin to check that only one wallet per community is allowed

**Comment by cammellos** [comment]
> Reference in new issue

(it would not prevent the issue instead it would mitigate it to some extend)

**Comment by saledjenic** [comment]
Seems there is no perfect solution for this issue since we cannot claim with 100% certainty that 2 addresses belong even to the same key, but even harder to different keys owned by the same person.


Does the following sound like a possible solution to you?

What Is we add another condition for joining a community, we can say that all addresses the one wants to reveal to the community _must be added to the profile showcase accounts list_. 

Knowing the fact that the only accounts that the Status user can add to the profile showcase accounts list are for sure, in 100% of cases, those which the Status user knows a private key for (cause no other way to add them except import them into the Status app).

Then a community just needs to check if revealed addresses belong to that list and then do the common check, checking signatures and what else is needed.

**Comment by John-44** [comment]
When you say:

> What Is we add another condition for joining a community, we can say that all addresses the one wants to reveal to the community must be added to the profile showcase accounts list.

We definitely don't want to force users to revel to other folks via the showcase the accounts that they wish to reveal to a community control node, this would be very bad for privacy.

However, 100% of a user's accounts are always in the profile showcase, just 'hidden' (not shared with anybody) by default.  If you mean that this includes accounts in the profile showcase that are set to 'hidden', then this would be ok, and would also include *all* accounts the user has (and doesn't include any 'viewed addresses').

**Comment by John-44** [comment]
> * if we want to prevent this, which is probably a product decision (@John-44 ) then an option would be to have the community admin to check that only one wallet per community is allowed

Do you mean one 'Wallet' or 'One account inside a Wallet'?  All Status profiles only have a single Wallet, that can contain many 'accounts' derived from many private keys.  I'm assuming that 'Account' is what you mean here, and we definitely don't want to limit users to only sharing a single account with a community control node.

**Comment by saledjenic** [comment]
> However, 100% of a user's accounts are always in the profile showcase, just 'hidden' (not shared with anybody) by default.

That's what I meant, but I didn't know that we were sharing all accounts, and that's why I said that "revealed accounts must be added to the profile showcase accounts list", but if so, then even better. 

That means the profile always contains info about accounts it owns (because those accounts can be added only if the profile knows their keys, which further means that it owns them).

Now the only we need to check when the profile is joining a community is if all revealed accounts belong to the list of profile's showcase accounts (hidden + exposed ones) and if so, then proceed with checking the requirements set by the community.

**Comment by iurimatias** [comment]
moving to 2.29 as we don't have space for it in 2.28 given current estimations

**Comment by jrainville** [comment]
Closing as outdated. Reopen if we really need it and/or when we have capacity
